File: PocketPrinter5890/Sources/PocketPrinter5890Kit/ResponseDecoder.swift
import Foundation

public enum ResponseDecoder {
    public static func decode(_ bytes: [UInt8], context: String?) -> String {
        let context = context?.lowercased() ?? ""

        // Acquittement ASCII observe sur FF01 apres une commande proprietaire
        // acceptee, par exemple le reglage de densite.
        if bytes == [0x4f, 0x4b] {
            return "OK: commande acceptee"
        }

        if bytes.count >= 3, bytes[0] == 0x02 {
            let suffix = bytes[2] != 0x00 ? " (flags 0x\(hex(bytes[2])))" : ""
            return "Batterie probable: \(bytes[1])%\(suffix)"
        }

        // Trame de credit de flux: `01 nn` signifie que l'imprimante peut
        // accepter nn paquets supplementaires. Ce n'est ni un acquittement
        // ni un statut papier, contrairement a ce que supposait la version
        // precedente de ce decodeur.
        if bytes.count == 2, bytes[0] == 0x01 {
            return "Credit de flux: +\(bytes[1]) paquet(s)"
        }

        if bytes.count >= 2, bytes[0] == 0x01 {
            return decodeStatusByte(bytes[1])
        }

        if context.contains("modele") || context.contains("model") {
            return ascii(bytes).map { "Modele: \($0)" } ?? "Modele: reponse non ASCII"
        }

        if context.contains("firmware") {
            return ascii(bytes).map { "Firmware: \($0)" } ?? "Firmware: reponse non ASCII"
        }

        if context.contains("serial") || context.contains("serie") {
            return ascii(bytes).map { "Numero de serie: \($0)" } ?? "Numero de serie: reponse non ASCII"
        }

        if context.contains("batterie") || context.contains("battery") {
            guard bytes.count >= 2 else { return "Batterie: reponse vide" }
            return "Batterie: \(bytes[1])%"
        }

        // Reponse a `10 FF 13`: delai d'extinction sur deux octets.
        if context.contains("extinction") || context.contains("shutdown") {
            guard bytes.count >= 2 else { return "Extinction auto: reponse vide" }
            let minutes = Int(bytes[bytes.count - 2]) << 8 | Int(bytes[bytes.count - 1])
            return minutes == 0
                ? "Extinction auto: desactivee"
                : "Extinction auto: \(minutes) min"
        }

        if context.contains("papier") || context.contains("paper") {
            switch bytes.last {
            case 0x00:
                return "Papier: present"
            case 0x04:
                return "Papier: capteur signale 0x04 (informatif, n'empeche pas d'imprimer)"
            case let value?:
                return "Papier: etat 0x\(String(format: "%02X", value))"
            case nil:
                return "Papier: reponse vide"
            }
        }

        if bytes == [0x00] {
            return "Statut: 0x00, papier present"
        }

        if let text = ascii(bytes), !text.isEmpty {
            return "ASCII: \(text)"
        }
        return ""
    }

    public static func looksLikeSolicitedResponse(_ bytes: [UInt8], context: String?) -> Bool {
        guard let context = context?.lowercased() else { return false }

        // Une trame `01 nn` est un credit de flux emis en continu pendant les
        // echanges: elle ne repond a aucune commande. La confondre avec une
        // reponse consommait le contexte et faisait lire `01 01` comme
        // « batterie 1 % » alors que la vraie reponse `00 62` suivait.
        if bytes.count == 2, bytes[0] == 0x01 { return false }

        if bytes == [0x4f, 0x4b] { return true }

        if context.contains("modele") || context.contains("model")
            || context.contains("firmware")
            || context.contains("serial")
            || context.contains("serie") {
            return ascii(bytes) != nil
        }

        if context.contains("batterie") || context.contains("battery") {
            // Deux formes observees: `02 64 00` emis spontanement, et
            // `00 62` en reponse a la requete. Dans les deux cas le
            // pourcentage est le deuxieme octet.
            guard bytes.count >= 2 else { return false }
            return bytes[0] == 0x02 || (bytes[1] > 0 && bytes[1] <= 100)
        }

        if context.contains("papier") || context.contains("paper") {
            return bytes.count == 1 || (bytes.count >= 2 && bytes[0] == 0x01)
        }

        return false
    }

    private static func ascii(_ bytes: [UInt8]) -> String? {
        let printable = bytes.filter { $0 >= 0x20 && $0 <= 0x7e }
        guard !printable.isEmpty else { return nil }
        return String(bytes: printable, encoding: .ascii)
    }

    private static func decodeStatusByte(_ status: UInt8) -> String {
        // Observation: pendant l'envoi d'un raster, l'imprimante emet un flot
        // continu de `01 01`. Il s'agit d'un acquittement de reception, pas
        // d'une erreur. Le statut `01 04` apparait a la connexion alors que
        // l'application officielle imprime sans probleme avec le meme papier:
        // il est donc informatif et ne doit jamais bloquer une impression.
        switch status {
        case 0x00:
            return "Etat 0x00: pret, papier present"
        case 0x01:
            return "Etat 0x01: reception ou impression en cours (normal)"
        case 0x04:
            return "Etat 0x04: capteur papier (informatif, n'empeche pas d'imprimer)"
        case 0x84:
            return "Etat 0x84: capteur papier + flag 0x80 (informatif)"
        default:
            var parts: [String] = []
            if status & 0x01 != 0 { parts.append("occupe") }
            if status & 0x04 != 0 { parts.append("capteur papier") }
            if status & 0x80 != 0 { parts.append("flag 0x80") }
            let known: UInt8 = 0x85
            let unknown = status & ~known
            if unknown != 0 { parts.append("flags inconnus 0x\(hex(unknown))") }
            if parts.isEmpty { parts.append("etat non documente") }
            return "Etat 0x\(hex(status)): \(parts.joined(separator: ", "))"
        }
    }

    private static func hex(_ byte: UInt8) -> String {
        String(format: "%02X", byte)
    }
}

File: android/kit/src/main/kotlin/fr/pocketprinter5890/kit/ResponseDecoder.kt
package fr.pocketprinter5890.kit

/** Formatage hexadecimal, pour les journaux. */
object Hex {
    fun encode(bytes: ByteArray): String =
        bytes.joinToString(" ") { "%02X".format(it.toInt() and 0xff) }
}

/**
 * Interprete les trames recues de l'imprimante.
 *
 * Le piege principal: une trame `01 nn` est un **credit de flux**, jamais une
 * reponse a une commande. La confondre avec une reponse fait lire `01 01`
 * comme « batterie 1 % » alors que la vraie reponse (`00 62`, 98 %) arrive
 * juste apres.
 */
object ResponseDecoder {

    fun decode(bytes: ByteArray, context: String?): String {
        val ctx = context?.lowercase() ?: ""
        val unsigned = bytes.map { it.toInt() and 0xff }

        // Trame de service emise spontanement en fin d'echange (`AA` suivi
        // d'un CRLF). Elle ne repond a aucune commande: la compter comme une
        // reponse decalait toutes les suivantes d'un cran, et son `0D` etait
        // lu comme « batterie 13 % ».
        if (isServiceFrame(bytes)) {
            return "Trame de service (fin d'echange)"
        }

        // Acquittement ASCII observe sur FF01 apres une commande proprietaire
        // acceptee, par exemple le reglage de densite.
        if (unsigned == listOf(0x4f, 0x4b)) {
            return "OK: commande acceptee"
        }

        if (unsigned.size >= 3 && unsigned[0] == 0x02) {
            val suffix = if (unsigned[2] != 0x00) " (flags 0x${hex(unsigned[2])})" else ""
            return "Batterie probable: ${unsigned[1]}%$suffix"
        }

        // Trame de credit de flux: `01 nn` signifie que l'imprimante peut
        // accepter nn paquets supplementaires. Ce n'est ni un acquittement ni
        // un statut papier.
        if (unsigned.size == 2 && unsigned[0] == 0x01) {
            return "Credit de flux: +${unsigned[1]} paquet(s)"
        }

        if (unsigned.size >= 2 && unsigned[0] == 0x01) {
            return decodeStatusByte(unsigned[1])
        }

        if (ctx.contains("modele") || ctx.contains("model")) {
            return ascii(bytes)?.let { "Modele: $it" } ?: "Modele: reponse non ASCII"
        }

        if (ctx.contains("firmware")) {
            return ascii(bytes)?.let { "Firmware: $it" } ?: "Firmware: reponse non ASCII"
        }

        if (ctx.contains("serial") || ctx.contains("serie")) {
            return ascii(bytes)?.let { "Numero de serie: $it" }
                ?: "Numero de serie: reponse non ASCII"
        }

        if (ctx.contains("batterie") || ctx.contains("battery")) {
            if (unsigned.size < 2) return "Batterie: reponse vide"
            return "Batterie: ${unsigned[1]}%"
        }

        // Reponse a `10 FF 13`: delai d'extinction sur deux octets.
        if (ctx.contains("extinction") || ctx.contains("shutdown")) {
            if (unsigned.size < 2) return "Extinction auto: reponse vide"
            val minutes = (unsigned[unsigned.size - 2] shl 8) or unsigned[unsigned.size - 1]
            return if (minutes == 0) {
                "Extinction auto: desactivee"
            } else {
                "Extinction auto: $minutes min"
            }
        }

        if (ctx.contains("papier") || ctx.contains("paper")) {
            return when (val last = unsigned.lastOrNull()) {
                null -> "Papier: reponse vide"
                0x00 -> "Papier: present"
                0x04 ->
                    "Papier: capteur signale 0x04 (informatif, n'empeche pas d'imprimer)"
                else -> "Papier: etat 0x${hex(last)}"
            }
        }

        if (unsigned == listOf(0x00)) {
            return "Statut: 0x00, papier present"
        }

        val text = ascii(bytes)
        if (!text.isNullOrEmpty()) {
            return "ASCII: $text"
        }
        return ""
    }

    /**
     * Determine si une trame repond a la commande en attente.
     *
     * Une trame `01 nn` est un credit emis en continu pendant les echanges:
     * elle ne repond a aucune commande. La compter comme une reponse
     * consommait le contexte et faisait lire `01 01` comme « batterie 1 % ».
     */
    fun looksLikeSolicitedResponse(bytes: ByteArray, context: String?): Boolean {
        val ctx = context?.lowercase() ?: return false
        val unsigned = bytes.map { it.toInt() and 0xff }

        if (unsigned.size == 2 && unsigned[0] == 0x01) return false
        if (isServiceFrame(bytes)) return false

        if (unsigned == listOf(0x4f, 0x4b)) return true

        if (ctx.contains("modele") || ctx.contains("model") ||
            ctx.contains("firmware") || ctx.contains("serial") || ctx.contains("serie")
        ) {
            return ascii(bytes) != null
        }

        if (ctx.contains("batterie") || ctx.contains("battery")) {
            // Deux formes observees: `02 64 00` emis spontanement, et `00 nn`
            // en reponse a la requete (`00 5E` = 94 % sur l'appareil de test).
            // Dans les deux cas le pourcentage est le deuxieme octet.
            if (unsigned.size < 2) return false
            if (unsigned[0] == 0x02) return true
            return unsigned[0] == 0x00 && unsigned[1] in 1..100
        }

        if (ctx.contains("papier") || ctx.contains("paper")) {
            // Un octet seul, ou `01 nn`. Une trame `00 nn` est une reponse
            // batterie et ne doit pas etre prise pour un etat papier.
            return unsigned.size == 1 || (unsigned.size >= 2 && unsigned[0] == 0x01)
        }

        return false
    }

    /**
     * Trame de service `AA 0D 0A`, emise spontanement en fin d'echange.
     *
     * Observee sur l'A2Y apres chaque commande. Ce n'est pas une reponse: la
     * traiter comme telle consomme le contexte en attente, et toutes les
     * reponses suivantes se retrouvent etiquetees avec la mauvaise commande.
     */
    fun isServiceFrame(bytes: ByteArray): Boolean {
        if (bytes.size != 3) return false
        return (bytes[0].toInt() and 0xff) == 0xaa &&
            (bytes[1].toInt() and 0xff) == 0x0d &&
            (bytes[2].toInt() and 0xff) == 0x0a
    }

    /**
     * Extrait un credit de flux d'une trame `01 nn`.
     *
     * @return le nombre de paquets supplementaires autorises, ou null si la
     *   trame n'est pas un credit.
     */
    fun creditFrame(bytes: ByteArray): Int? {
        if (bytes.size != 2) return null
        if ((bytes[0].toInt() and 0xff) != 0x01) return null
        return bytes[1].toInt() and 0xff
    }

    private fun ascii(bytes: ByteArray): String? {
        val printable = bytes.filter { (it.toInt() and 0xff) in 0x20..0x7e }
        if (printable.isEmpty()) return null
        return String(printable.toByteArray(), Charsets.US_ASCII)
    }

    private fun decodeStatusByte(status: Int): String {
        // Observation: pendant l'envoi d'un raster, l'imprimante emet un flot
        // continu de `01 01`. Il s'agit d'un acquittement de reception, pas
        // d'une erreur. Le statut `01 04` apparait a la connexion alors que
        // l'application officielle imprime sans probleme avec le meme papier:
        // il est donc informatif et ne doit jamais bloquer une impression.
        return when (status) {
            0x00 -> "Etat 0x00: pret, papier present"
            0x01 -> "Etat 0x01: reception ou impression en cours (normal)"
            0x04 -> "Etat 0x04: capteur papier (informatif, n'empeche pas d'imprimer)"
            0x84 -> "Etat 0x84: capteur papier + flag 0x80 (informatif)"
            else -> {
                val parts = mutableListOf<String>()
                if (status and 0x01 != 0) parts.add("occupe")
                if (status and 0x04 != 0) parts.add("capteur papier")
                if (status and 0x80 != 0) parts.add("flag 0x80")
                val known = 0x85
                val unknown = status and known.inv() and 0xff
                if (unknown != 0) parts.add("flags inconnus 0x${hex(unknown)}")
                if (parts.isEmpty()) parts.add("etat non documente")
                "Etat 0x${hex(status)}: ${parts.joinToString(", ")}"
            }
        }
    }

    private fun hex(byte: Int): String = "%02X".format(byte and 0xff)
}
